File: src/podcast/routers/feed.py
from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import Response
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from podcast.database import get_db
from podcast.models import User
from podcast.services.feed import generate_feed

router = APIRouter()


@router.get("/feed/{feed_token}.xml", name="feed")
async def feed(feed_token: str, db: AsyncSession = Depends(get_db)):
    """Serve a user's podcast RSS feed by their unique feed token."""
    result = await db.execute(
        select(User).where(User.feed_token == feed_token, User.enabled == True)
    )
    user = result.scalar_one_or_none()
    if not user:
        raise HTTPException(status_code=404, detail="Feed not found")

    xml = await generate_feed(db, user)
    return Response(content=xml, media_type="application/rss+xml; charset=utf-8")

File: src/routes/feed.ts
import type { FastifyInstance } from "fastify";
import { eq, and } from "drizzle-orm";
import { db } from "../database.js";
import { users } from "../schema.js";
import { generateFeed } from "../services/feed.js";

export async function feedRoutes(app: FastifyInstance) {
  app.get<{ Params: { feedToken: string } }>(
    "/feed/:feedToken.xml",
    async (request, reply) => {
      const { feedToken } = request.params;
      const [user] = await db
        .select()
        .from(users)
        .where(and(eq(users.feedToken, feedToken), eq(users.enabled, true)))
        .limit(1);

      if (!user) {
        return reply.code(404).send({ error: "Feed not found" });
      }

      const xml = await generateFeed(user);
      return reply
        .header("Content-Type", "application/rss+xml; charset=utf-8")
        .send(xml);
    }
  );
}
